File: leetcode1106/solution.go
/*
Solution: Recursion

Time: O(n) | Space: O(n)

Runtime: 0 ms | 100.00%
Memory: 4.63 MB | 18.18%

- n: length of 'expression'
*/

func parseBoolExpr(expression string) bool {
	current := len(expression) - 1

	return recursive(expression, &current)
}

func recursive(exp string, current *int) bool {
	stack := []bool{}

	for *current >= 0 {
		switch exp[*current] {
		case 'f':
			stack = append(stack, false)
			*current -= 1
		case 't':
			stack = append(stack, true)
			*current -= 1
		case '|':
			res := stack[0]

			for i := 1; i < len(stack); i++ {
				res = res || stack[i]
			}

			return res
		case '&':
			res := stack[0]

			for i := 1; i < len(stack); i++ {
				res = res && stack[i]
			}

			return res
		case '!':
			res := stack[0]

			return !res
		case ')':
			*current -= 1
			stack = append(stack, recursive(exp, current))
			*current -= 1
		default:
			*current -= 1
		}
	}

	return stack[0]
}

File: leetcode1106/solution.rs
/*
Solution: Recursion

Time: O(n) | Space: O(n)

Runtime: 0 ms | 100.00%
Memory: 2.19 MB | 28.57%

- n: length of expression
*/

impl Solution {
    pub fn parse_bool_expr(expression: String) -> bool {
        let mut chars: Vec<char> = expression.chars().collect::<Vec<char>>();

        Self::recursive(&mut chars)
    }

    fn recursive(chars: &mut Vec<char>) -> bool {
        let mut stack: Vec<bool> = Vec::new();

        while let Some(c) = chars.pop() {
            match c {
                'f' => stack.push(false),
                't' => stack.push(true),
                '|' => {
                    let mut res: bool = stack.pop().unwrap();

                    while let Some(b) = stack.pop() {
                        res |= b;
                    }

                    return res;
                },
                '&' => {
                    let mut res: bool = stack.pop().unwrap();

                    while let Some(b) = stack.pop() {
                        res &= b;
                    }

                    return res;
                },
                '!' => {
                    let mut res: bool = stack.pop().unwrap();
                    
                    return !res;
                },
                ')' => {
                    stack.push(Self::recursive(chars));
                }
                _ => ()
            }
        }

        stack.pop().unwrap()
    }
}